docs/TRACEBACKS.md

```diff
@@ -0,0 +1,198 @@
+When a Python program encounters an unhandled exception, it will print the exception message and a Traceback.
+The Traceback will show where the exception was raised and what functions were called leading up to it.
+
+## How to Read a Traceback
+
+`ValueError` is a common exception.
+
+Following is an example of `ValueError` resulting from trying to assign two variables on the left from only one value on the right:
+
+```python
+>>> first, second = [1]
+Traceback (most recent call last):
+File <stdin>, line 1, in <module>
+    first, second = [1]
+ValueError: not enough values to unpack (expected 2, got 1)
+
+```
+
+It says that the most recent call is last, so the place to start is with the exception at the bottom of the trace.
+From there it can be traced back to the start.
+If we place the offending line in a function and then call the function we see a longer trace:
+
+```python
+>>> def my_func():
+...     first, second = [1]
+... 
+>>> my_func()
+Traceback (most recent call last):
+  File <stdin>, line 5, in <module>
+    my_func()
+  File <stdin>, line 2, in my_func
+    first, second = [1]
+ValueError: not enough values to unpack (expected 2, got 1)
+
+```
+
+Working backwards from the bottom, we see that the call where the exception happened is on line 2 in `my_func`.
+We got there by calling `my_func` on line 5.
+
+## Using the `print` function
+
+Sometimes an error is not being raised, but a value is not what is expected.
+This can be especially perplexing if the value is the result of a chain of calculations.
+In such a situation it can be helpful to look at the value at each step to see which step is the one that isn't behaving as expected.
+The [print][print] function can be used for printing the value to the console.
+Following is an example of a function that doesn't return the value expected
+
+```python
+# the intent is to only pass an int to this function and get an int back
+def halve_and_quadruple(num):
+    return (num / 2) * 4
+```
+
+When the function is passed `5`, the expected value is `8`, but it it returns `10.0`.
+To troubleshoot, the calculating is broken up so that the value can be inspected at every step.
+
+
+```python
+# the intent is to only pass an int to this function and get an int back
+def halve_and_quadruple(num):
+    # verify the number in is what is expected
+    # prints 5
+    print(num)
+    # we want the int divided by an int to be an int
+    # but this prints 2.5! We've found our mistake.
+    print(num / 2)
+    # this makes sense, since 2.5 x 4 = 10.0
+    print((num / 2) * 4)
+    return (num / 2) * 4
+
+```
+
+What the `print` calls revealed is that we used `/` when we should have used `//`, the [floor divison operator][floor divison operator].
+
+## Logging
+
+[Logging][logging] can be used similarly to `print`, but it is more powerful.
+Logging can print to the console, but it can also be configured to print to a file.
```

Comment:
But we _can_ `print` to a file, so this comparison and reason why logging is "more powerful" is not really an honest reasoning.

Can we maybe come up with a better reason why we should use the logging facility?  Perhaps rotation of logs is built in, or some other feature of logging that `print` does not consider doing.

Reply:
I'm probably a bit out of my depth here, but I'll try to work in something from [this](https://stackoverflow.com/questions/6918493/in-python-why-use-logging-instead-of-print).

Reply:
Things like loggers generally flush on any output, to ensure that logging is done, unlike functions that output to the terminal.  The terminal may also have controls over buffer to standard out and standard error.  Logging facilities often, also, try to ensure no loss of data, so may have safeguards to write even after most "catastrophic" failures in the program runtime.

These need should be confirmed for the language, but would be reasons why I would suggest logging over the normal way of output in a language, often via `print`.

Reply:
Okay, I added  a bit to better and more accurately give reasons for logging.

Reply:
Downloading the latest changes to review… Will check in about 30, as Isaac is busy here as well, recently.